### PR TITLE
#458 Automate Docker Requirements Workflow for PR Creation and Review Requests

### LINKED ISSUE #458: Automating Workflow for Docker Requirement Fixes
Currently, the `docker-requirements / update (pull_request)` workflow is configured to be triggered manually. When activated, it opens a separate PR to fix Docker requirements. While this ensures updates are controlled, it requires manual intervention, which can delay necessary fixes.

Maybe we can configure the workflow to be automatically triggered on every push to the main branch, creating a PR if Docker requirements need fixing. cc: @pmeier

### PR DESCRIPTION
**Pull request by arjxn-py**
Should Fix #458 

This PR updates the Docker requirements workflow to:

- Automatically create PRs to update requirements-docker.lock when changes are detected.
- Dynamically request reviews from the author and the person who merged the last PR into main.

I've also tested this updated workflow a number of times on my fork to- 
- Here's the [test workflow run](https://github.com/arjxn-py/ragna/actions/runs/10049115488/job/27774541367#step:6:1)
- Here's [the automatically created PR](https://github.com/arjxn-py/ragna/pull/6) when I bumped down `panel==1.4.2` on my fork for testing.

### CODE REVIEW

**Comment by arjxn-py** [comment]
We should also set a repository secret with name `GH_TOKEN` that would most likely be a fine grained access token. cc: @pmeier

File: .github/workflows/update-docker-requirements.yml [hunk 1]
```diff
@@ -1,11 +1,9 @@
...
 
 on:
-  pull_request:
-    paths:
-      - ".github/workflows/update-docker-requirements.yml"
-      - "pyproject.toml"
-      - "requirements-docker.lock"
+  push:
+    branches:
+      - main
```
Comment (arjxn-py) [inline .github/workflows/update-docker-requirements.yml:6]:
```suggestion
      - main
    paths:
      - ".github/workflows/update-docker-requirements.yml"
      - "pyproject.toml"
      - "requirements-docker.lock"
```

If you'd only like to run this if these files are changes, feel free to commit this suggestion.

File: .github/workflows/update-docker-requirements.yml [hunk 2]
```diff
@@ -42,10 +43,22 @@ jobs:
           NEEDS_UPDATE=$(git diff --quiet && echo 'false' || echo 'true')
           echo "needs-update=${NEEDS_UPDATE}" | tee --append $GITHUB_OUTPUT
 
+      - name: Get last merged PR info
+        id: pr_info
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          LAST_MERGED_PR=$(gh pr list --repo $GITHUB_REPOSITORY --state closed --json number --jq '.[0].number')
+          echo "Last merged PR number: $LAST_MERGED_PR"
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:52]: This logic is flawed unfortunately. Just because we have the last merged PR, it doesn't have to be merged into `main`. Imagine merging a PR into a feature branch.

Instead, we need to query the API for the [PR associated with the commit](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#list-pull-requests-associated-with-a-commit) we are currently running the workflow on. It should be something like

```bash
RESPONSE=$(gh api \
  -H 'Accept: application/vnd.github+json' \
  -H 'X-GitHub-Api-Version: 2022-11-28' \
  /repos/{owner}/{repo}/commits/${{ github.sha }}/pulls)
if [ $? -ne 0 ]; then
  echo 'has-pr=false' | tee --append $GITHUB_OUTPUT
  exit 0
fi

echo 'has-pr=true' | tee --append $GITHUB_OUTPUT
PR_NUMBER=$(jq '.[0].number' <<< "${RESPONSE}")
```

We have to adjust the body of the PR that we open based on `${{ ${{ steps.pr-info.outputs.has-pr }} }}`.

File: .github/workflows/update-docker-requirements.yml [hunk 3]
```diff
@@ -42,10 +43,22 @@ jobs:
           NEEDS_UPDATE=$(git diff --quiet && echo 'false' || echo 'true')
           echo "needs-update=${NEEDS_UPDATE}" | tee --append $GITHUB_OUTPUT
 
+      - name: Get last merged PR info
+        id: pr_info
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:49]: Do we really need that? Everything we do here should be public information.

File: .github/workflows/update-docker-requirements.yml [hunk 4]
```diff
@@ -42,10 +43,22 @@ jobs:
           NEEDS_UPDATE=$(git diff --quiet && echo 'false' || echo 'true')
           echo "needs-update=${NEEDS_UPDATE}" | tee --append $GITHUB_OUTPUT
 
+      - name: Get last merged PR info
+        id: pr_info
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:47]: Let's name this
 
```suggestion
        id: pr-info
```

as we do everywhere else in this repo as well.

File: .github/workflows/update-docker-requirements.yml [hunk 5]
```diff
@@ -42,10 +43,22 @@ jobs:
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          LAST_MERGED_PR=$(gh pr list --repo $GITHUB_REPOSITORY --state closed --json number --jq '.[0].number')
+          echo "Last merged PR number: $LAST_MERGED_PR"
+          PR_INFO=$(gh pr view $LAST_MERGED_PR --repo $GITHUB_REPOSITORY --json author,mergedBy --jq '{author: .author.login, mergedBy: .mergedBy.login}')
+          echo "PR Info: $PR_INFO"
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:54]: That is just duplicated information with the stuff below.

File: .github/workflows/update-docker-requirements.yml [hunk 6]
```diff
@@ -42,10 +43,22 @@ jobs:
...
+          echo "pr_author=$(echo $PR_INFO | jq -r .author)" >> $GITHUB_ENV
+          echo "pr_mergedBy=$(echo $PR_INFO | jq -r .mergedBy)" >> $GITHUB_ENV
+          echo "PR Author: $(echo $PR_INFO | jq -r .author)"
+          echo "PR MergedBy: $(echo $PR_INFO | jq -r .mergedBy)"
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:58]: 1. Let's not assign this to environment variable if its just arbitrary information rather than something another tool consumes automatically.
2. Instead `echo`ing the value separately, let's use `tee`

```suggestion
          echo "author=$(echo ${PR_INFO} | jq -r .author)" | tee --append $GITHUB_OUTPUT
          echo "merger=$(echo ${PR_INFO} | jq -r .merger)" | tee --append $GITHUB_OUTPUT
```

With this, you can now access the information with `${{ steps.pr-info.outputs.author }}`.

File: .github/workflows/update-docker-requirements.yml [hunk 7]
```diff
@@ -42,10 +43,22 @@ jobs:
...
           echo "needs-update=${NEEDS_UPDATE}" | tee --append $GITHUB_OUTPUT
 
+      - name: Get last merged PR info
+        id: pr_info
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          LAST_MERGED_PR=$(gh pr list --repo $GITHUB_REPOSITORY --state closed --json number --jq '.[0].number')
+          echo "Last merged PR number: $LAST_MERGED_PR"
+          PR_INFO=$(gh pr view $LAST_MERGED_PR --repo $GITHUB_REPOSITORY --json author,mergedBy --jq '{author: .author.login, mergedBy: .mergedBy.login}')
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:53]: Nit: simplify

```suggestion
          PR_INFO=$(gh pr view $LAST_MERGED_PR --repo $GITHUB_REPOSITORY --json author,mergedBy --jq '{author: .author.login, merger: .mergedBy.login}')
```

File: .github/workflows/update-docker-requirements.yml [hunk 8]
```diff
@@ -56,10 +69,8 @@ jobs:
           # prettier-ignore
           body:
             https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
-          reviewers: ${{ github.actor }}
+          reviewers: ${{ env.pr_author }},${{ env.pr_mergedBy }}
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:72]: We can leave the merger here, but I don't know whether we can put an arbitrary user as reviewer. For example, I cannot ask for your review on a PR. We should just ping the author in the body.

File: .github/workflows/update-docker-requirements.yml [hunk 9]
```diff
@@ -56,10 +69,8 @@ jobs:
...
 
       - name: Show diff
-        # prettier-ignore
-        if:
-          ${{ steps.update.outputs.needs-update && github.event_name != 'workflow_dispatch' }}
+        if: ${{ steps.update.outputs.needs-update }}
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:75]: This whole step can be removed as we either get PR or we don't. No one will look at the action run anyway.

**Review by pmeier — COMMENTED** [review]
Thanks Arjun. I left a bunch of comments. We also need to think of a mechanism to surface a failure of this workflow somehow, because if we never get PRs that could also mean that the workflow is broken. So maybe use an action that opens an issue on failures?

File: .github/workflows/update-docker-requirements.yml (hunk 2 above)
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: > Thoughts?

Your suggestion makes sense, Thanks. 
I'm on it to refine it the same way.

File: .github/workflows/update-docker-requirements.yml [hunk 10]
```diff
@@ -42,24 +47,40 @@ jobs:
...
+        id: pr-check
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          PR_NUMBER=$(gh api \
+            -H 'Accept: application/vnd.github+json' \
+            /repos/${{ github.repository }}/commits/${{ github.sha }}/pulls \
+            --jq '.[0].number' || echo "none")
+          if [ "${PR_NUMBER}" != "none" ]; then
+            HAS_PR='true'
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:60]: I guess we don't even need this information and can just check the merger or author for "none". Or am I missing something? If need, let's remove it.

File: .github/workflows/update-docker-requirements.yml [hunk 11]
```diff
@@ -42,24 +47,40 @@ jobs:
...
 
+      - name: Check if commit is associated with a PR
+        id: pr-check
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          PR_NUMBER=$(gh api \
+            -H 'Accept: application/vnd.github+json' \
+            /repos/${{ github.repository }}/commits/${{ github.sha }}/pulls \
+            --jq '.[0].number' || echo "none")
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:58]: Why the change to defaulting to "none" here instead of relying on the exit code?

File: .github/workflows/update-docker-requirements.yml [hunk 12]
```diff
@@ -42,24 +47,40 @@ jobs:
...
-          reviewers: ${{ github.actor }}
-
-      - name: Show diff
-        # prettier-ignore
-        if:
-          ${{ steps.update.outputs.needs-update && github.event_name != 'workflow_dispatch' }}
-        run: git diff --exit-code
+          body: |
+            Automatic update of requirements-docker.lock.
+            [Workflow run details](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})
```
Comment (pmeier) [inline .github/workflows/update-docker-requirements.yml:85]: Currently the author information is not used. We should just add the equivalent of `cc @${{ author }}` to the body if we have an author.

File: .github/workflows/update-docker-requirements.yml (hunk 3 above)
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml:50]: I guess yes, as when i tried removing that, it throws this error - [https://github.com/arjxn-py/ragna/actions/runs/10141546877/job/28039000468#step:6:26](https://github.com/arjxn-py/ragna/actions/runs/10141546877/job/28039000468#step:6:26)

File: .github/workflows/update-docker-requirements.yml (hunk 5 above)
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: Removed

File: .github/workflows/update-docker-requirements.yml (hunk 12 above)
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml:85]: That's done, cc: @author if author!=None

File: .github/workflows/update-docker-requirements.yml (hunk 10 above)
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: Removed it, yet to test it. 
When testing this branch on my fork it fails because it cannot find #459 PR on the fork.

Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: here is a test run - [https://github.com/arjxn-py/ragna/actions/runs/10183118614/job/28167297481](https://github.com/arjxn-py/ragna/actions/runs/10183118614/job/28167297481)

**Comment by arjxn-py** [comment]
Here'a a better test run - [https://github.com/arjxn-py/ragna/actions/runs/10183204935/job/28167581020](https://github.com/arjxn-py/ragna/actions/runs/10183204935/job/28167581020)

& the sample PR that's been opened as a result - [https://github.com/arjxn-py/ragna/pull/10](https://github.com/arjxn-py/ragna/pull/10)

**Comment by arjxn-py** [comment]
> We also need to think of a mechanism to surface a failure of this workflow somehow, because if we never get PRs that could also mean that the workflow is broken. So maybe use an action that opens an issue on failures?

Question: if there are No PR's, there's no update on dependencies. Does `requirements-docker.lock` still needs updates in that case?

**Review by pmeier — COMMENTED** [review]
Looking pretty good Arjun! A couple more comments inline.

**Comment by pmeier** [comment]
> if there are No PR's, there's no update on dependencies. Does `requirements-docker.lock` still needs updates in that case?

No. If we (sadly) ever get to the point where there are no commits on this repo anymore, there is also no need to update docker requirements. If that somehow is needed for some reason, you can also do `git checkout main && git commit -m 'trigger CI' --allow-empty && git push`.

File: .github/workflows/update-docker-requirements.yml [hunk 13]
```diff
@@ -42,24 +45,73 @@ jobs:
...
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+        run: |
+          PR_NUMBER=$(gh api \
+            -H 'Accept: application/vnd.github+json' \
+            /repos/${{ github.repository }}/commits/${{ github.sha }}/pulls \
+            --jq '.[0].number')
+          if [ $? -eq 0 ]; then
+            PR_INFO=$(gh pr view ${PR_NUMBER} --json author,mergedBy --jq '{author: .author.login, merger: .mergedBy.login}')
+
+            if [ $? -eq 0 ]; then
```
[the remark this answers is not in the document]
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: Thanks for pointing this out. I guess i overdid it 😅 
Just simplified the logic

File: .github/workflows/update-docker-requirements.yml [hunk 14]
```diff
@@ -42,24 +45,73 @@ jobs:
...
 
-      - name: Show diff
-        # prettier-ignore
-        if:
-          ${{ steps.update.outputs.needs-update && github.event_name != 'workflow_dispatch' }}
-        run: git diff --exit-code
+  notify-failure:
+    runs-on: ubuntu-latest
+    needs: update
+    if: failure()
```
[the remark this answers is not in the document]
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: Done

File: .github/workflows/update-docker-requirements.yml [hunk 15]
```diff
@@ -42,24 +45,73 @@ jobs:
...
+    steps:
+      - name: Create failure issue
+        uses: actions/github-script@v6
```
[the remark this answers is not in the document]
Comment (arjxn-py) [reply .github/workflows/update-docker-requirements.yml]: Thanks for suggesting, I switched to this only.
I didn't actually looked into for a previous existing action & just started implementing. Something for me to keep in mind from next time.

**Comment by arjxn-py** [comment]
> No. If we (sadly) ever get to the point where there are no commits on this repo anymore, there is also no need to update docker requirements. If that somehow is needed for some reason, you can also do `git checkout main && git commit -m 'trigger CI' --allow-empty && git push`.

I don't think we need `git checkout main && git commit -m 'trigger CI' --allow-empty && git push` or it'd help as we have certain files that trigger this workflow. But it should be a nice thing to configure something like `dependabot`, I believe this workflow would compliment it nicely. And also keep the project updated.

**Review by pmeier — APPROVED** [review]
LGTM, thanks Arjun!
